# 1. Data import and Data Cleaning


## Team Best Practice (Data Science)

<details>
<summary>Click to expand: Team Best Practice</summary>

**Recommended setup**

### 👥 Each team member

Works on their own branch:

```bash
git checkout -b florian-analysis
```

---

### 🚀 Push your work

```bash
git push origin florian-analysis
```

---

### 🔀 Merge workflow (on GitHub)

- Open a Pull Request (PR)
- Team reviews changes
- Merge into `main`

👉 This keeps the project clean and avoids conflicts.

---

## ⚠️ Important for Colab

Colab is **not a version control system**:

- Changes are only stored in the current session  
- If the runtime restarts → work is lost  

---

## 💡 Therefore always:

> **Colab + GitHub = proper saving & version control**

## 1.1 Repository in GitHub

In [ ]:
# clone GitHub repository
!rm -rf Eniac_project_part2                                       # only if there is an error
!git clone https://github.com/FlorianHast/Eniac_project_part2.git


# lists the files in the folder 'data'
!ls Eniac_project_part2/data/

# used to shorten the path to the current repository
# %cd REPO

# for later:
# git add .
# git commit -m "Colab update"
# git push

Cloning into 'Eniac_project_part2'...
remote: Enumerating objects: 67, done.
remote: Counting objects: 100% (67/67), done.
remote: Compressing objects: 100% (62/62), done.
remote: Total 67 (delta 27), reused 0 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (67/67), 6.09 MiB | 5.01 MiB/s, done.
Resolving deltas: 100% (27/27), done.
brands.csv  orderlines.csv  orders.csv	products.csv


## 1.2 Import Data from GitHub

In [77]:
import pandas as pd

# if something goes wrong with your DataFrame you can reset them here by running this code block!

# orders.csv
orders_df = pd.read_csv('Eniac_project_part2/data/orders.csv')

# orderlines.csv
orderlines_df = pd.read_csv('Eniac_project_part2/data/orderlines.csv')

# products.csv
products_df = pd.read_csv('Eniac_project_part2/data/products.csv')

# brands.csv
brands_df = pd.read_csv('Eniac_project_part2/data/brands.csv')

## 1.3 Overview RawData

In [ ]:
orders_df.head(10)

,order_id,created_date,total_paid,state
0,241319,2017-01-02 13:35:40,44.99,Cancelled
1,241423,2017-11-06 13:10:02,136.15,Completed
2,242832,2017-12-31 17:40:03,15.76,Completed
3,243330,2017-02-16 10:59:38,84.98,Completed
4,243784,2017-11-24 13:35:19,157.86,Cancelled
5,245275,2017-06-28 11:35:37,149.00,Completed
6,245595,2017-01-21 12:52:47,112.97,Completed
7,245851,2017-04-04 20:58:21,79.99,Pending
8,245941,2017-01-01 10:32:23,183.52,Completed
9,246018,2017-02-13 23:02:57,211.95,Completed


In [ ]:
orderlines_df.head(10)

,id,id_order,product_id,product_quantity,sku,unit_price,date
0,1119109,299539,0,1,OTT0133,18.99,2017-01-01 00:07:19
1,1119110,299540,0,1,LGE0043,399.00,2017-01-01 00:19:45
2,1119111,299541,0,1,PAR0071,474.05,2017-01-01 00:20:57
3,1119112,299542,0,1,WDT0315,68.39,2017-01-01 00:51:40
4,1119113,299543,0,1,JBL0104,23.74,2017-01-01 01:06:38
5,1119114,295310,0,10,WDT0249,231.79,2017-01-01 01:14:27
6,1119115,299544,0,1,APP1582,1.137.99,2017-01-01 01:17:21
7,1119116,299545,0,1,OWC0100,47.49,2017-01-01 01:46:16
8,1119119,299546,0,1,IOT0014,18.99,2017-01-01 01:50:34
9,1119120,295347,0,1,APP0700,72.19,2017-01-01 01:54:11


In [ ]:
products_df.head(10)

,sku,name,desc,price,promo_price,in_stock,type
0,RAI0007,Silver Rain Design mStand Support,Aluminum support compatible with all MacBook,59.99,499.899,1,8696
1,APP0023,Apple Mac Keyboard Keypad Spanish,USB ultrathin keyboard Apple Mac Spanish.,59,589.996,0,13855401
2,APP0025,Mighty Mouse Apple Mouse for Mac,mouse Apple USB cable.,59,569.898,0,1387
3,APP0072,Apple Dock to USB Cable iPhone and iPod white,IPhone dock and USB Cable Apple iPod.,25,229.997,0,1230
4,KIN0007,Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM,2GB RAM Mac mini and iMac (2006/07) MacBook Pr...,34.99,31.99,1,1364
5,APP0073,Apple Composite AV Cable iPhone and iPod white,IPhone and iPod AV Cable Dock to Composite Video.,45,420.003,0,1230
6,KIN0008,Mac Memory Kingston 1GB 667MHz DDR2 SO-DIMM,1GB RAM Mac mini and iMac (2006/07) MacBook Pr...,18.99,146.471,0,1364
7,KIN0009,Mac Memory Kingston 2GB 800MHz DDR2 SO-DIMM,2GB RAM iMac with Intel Core 2 Duo (Penryn).,36.99,274.694,0,1364
8,KIN0001-2,Mac memory Kingston 4GB (2x2GB) 667MHz DDR2 SO...,RAM 4GB (2x2GB) Mac mini and iMac (2006/07) Ma...,74,669.904,0,1364
9,APP0100,Apple Adapter Mini Display Port to VGA,Adapter Mini Display Port to VGA MacBook and M...,35,330.003,0,1325


In [ ]:
brands_df.head(10)

,short,long
0,8MO,8Mobility
1,ACM,Acme
2,ADN,Adonit
3,AII,Aiino
4,AKI,Akitio
5,ALL,Allocacoc
6,AP2,Apple
7,APP,Apple
8,BAN,Band&Strap
9,BEA,Beats


# 2. Data Cleaning

## 2.1 orders.csv

In [78]:
orders_df.info()

# drop the rows with NaN values in 'total_paid':
orders_df = orders_df.dropna(subset=['total_paid'])

# check if rows with NaN values actually got dropped
NaN_remaining = orders_df['total_paid'].isna().sum()

print(f"There are {NaN_remaining} 'NaN' values remaining")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226909 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      226909 non-null  int64  
 1   created_date  226909 non-null  object 
 2   total_paid    226904 non-null  float64
 3   state         226909 non-null  object 
dtypes: float64(1), int64(1), object(2)
memory usage: 6.9+ MB
There are 0 'NaN' values remaining


In [79]:
# check for duplicates
orders_df.duplicated().sum()

np.int64(0)

In [80]:
# format 'create_date' as datetime
orders_df['created_date'] = pd.to_datetime(orders_df['created_date'])

orders_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 226904 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   order_id      226904 non-null  int64         
 1   created_date  226904 non-null  datetime64[ns]
 2   total_paid    226904 non-null  float64       
 3   state         226904 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(1)
memory usage: 8.7+ MB


In [81]:
# check for range in 'total paid'
orders_df.sort_values(['total_paid'])

# check for range in 'created_date'
orders_df.sort_values(['created_date'])

# check if there are more states than the 5 described in LMS
orders_df['state'].unique()

array(['Cancelled', 'Completed', 'Pending', 'Shopping Basket',
       'Place Order'], dtype=object)

Problem with the data: <br>
Orders.csv contains<br>
* orders with a 'total_paid' of 0,- € with 'place order' state (might be due to discounts?)<br>
* and orders with +200.000,- € paid(!) with 'Shopping Basket' state (if in basket, it´s not paid yet...)

## 2.2 orderlines.csv

In [82]:
# drop the outdated, unused 'product_id'
orderlines_df = orderlines_df.drop(['product_id'], axis=1)

In [83]:
# format 'date' as datetime
orderlines_df['date'] = pd.to_datetime(orderlines_df['date'])

In [91]:
# replace '.' with '' if it is not a decimal point
orderlines_df['unit_price'] = (orderlines_df['unit_price'].str.replace(r'\.(?=.*\.)', '', regex=True))

# format 'unit_price' as float
orderlines_df['unit_price'] = orderlines_df['unit_price'].astype(float)

# check if there are commas -> NO commas!
# orderlines_df['unit_price'].str.contains(',').sum()

In [103]:
# check if the 'unit_price' has still troublesome values
# orderlines_df.sort_values('unit_price').tail(5)
# orderlines_df.loc[orderlines_df['unit_price'] > 1100].sort_values('unit_price').tail(5)

orderlines_df.describe()

,id,id_order,product_quantity,unit_price,date
count,2.939830e+05,293983.000000,293983.000000,293983.000000,293983
mean,1.397918e+06,419999.116544,1.121126,409.855192,2017-09-19 03:19:26.305779712
min,1.119109e+06,241319.000000,1.000000,-119.000000,2017-01-01 00:07:19
25%,1.262542e+06,362258.500000,1.000000,35.990000,2017-06-06 16:20:34.500000
50%,1.406940e+06,425956.000000,1.000000,92.990000,2017-11-13 21:13:53
75%,1.531322e+06,478657.000000,1.000000,339.580000,2018-01-02 04:47:03
max,1.650203e+06,527401.000000,999.000000,159989.830000,2018-03-14 13:58:36
std,1.530096e+05,66344.486479,3.396569,837.065461,NaN


Problem with the data: <br>
Orderlines.csv contains<br>
* orders with a 'unit_price' of 0,- € and even a negative value (-119,-, might be due to discounts or typo?)<br>
* and unit_price over 999.99 € are sorted in the 1, 2, ... and so forth, because<br>
  * there´s a issue with the thousands delimiter (e.g. 1.137.99)<br>
  * datatype as string (maybe replace all ‘.’ (except decimal) with ‘’ (nothing) <br>
  * are there also commas in the mix? -> NO
* highest values in unit_price over 10000.- €, max is even over 150000.- €!

## 2.3 products.csv

In [108]:
products_df.info()
products_df.columns()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19326 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          19326 non-null  object
 1   name         19326 non-null  object
 2   desc         19319 non-null  object
 3   price        19280 non-null  object
 4   promo_price  19326 non-null  object
 5   in_stock     19326 non-null  int64 
 6   type         19276 non-null  object
dtypes: int64(1), object(6)
memory usage: 1.0+ MB


TypeError: 'Index' object is not callable